### DISCUSSION TITLE
Understanding validators order

### DISCUSSION THREAD

**Discussion by ivan94fi** [pydantic/pydantic#4270] [Question] [answered]
Hi I opened the issue [4268](https://github.com/samuelcolvin/pydantic/issues/4268), which was marked by @samuelcolvin as expected behaviour and not an issue. However I am still having problems understanding the logic. The code snippet I reported the bug for is the following:
```python
from pydantic import BaseSettings, validator

class Config(BaseSettings):
    input_resolution: int
    output_resolution: int
    scale_factor: int = None

    @validator("scale_factor")
    def compute_scale_factor(cls, v, values):
        if v is not None:
            raise ValueError("scale_factor cannot be passed, it is computed automatically")
        out_res = values["output_resolution"]
        in_res = values["input_resolution"]

        scale_factor = out_res / in_res
        if not scale_factor.is_integer():
            raise ValueError("Scale factor must be integer.")

        return int(scale_factor)

    class Config:
        validate_assignment = True
        validate_all = True
        extra = "forbid"
        env_file_encoding = "utf-8"
        env_nested_delimiter = "__"

c = Config()
```

If executed, the code fails with a `KeyError` inside `scale_factor` validator, while I would expect it to fail even before entering the scale factor validator, as `input_resolution` and `output_resolution` are required fields and they are not passed in the constructor.

The suggestion from @samuelcolvin was to catch the `KeyError` and return.

So, where is this documented and why this happens?

Thank you!

**Comment by samuelcolvin** [comment, marked as answer]
Because we run validation on all fields, even if one fails.

Otherwise, we could only ever return details about the first failed validation.

**Reply by samuelcolvin** [reply]
https://pydantic-docs.helpmanual.io/usage/validators/

> If validation fails on another field (or that field is missing) it will not be included in values, hence if 'password1' in values and ... in this example.

**Reply by ivan94fi** [reply]
Wow you are too fast for me. I just found the solution and wanted to answer my own question but you were faster. Thank you anyway.

**Comment by ivan94fi** [comment]
I just realized my mistake, but stil posted the question just to answer it, hoping it will be useful for someone else.

The problem was that I was not considering how Pydantic handles validation errors: what happens is that validation is carried out for every field and only at the end the validation errors are reported. But in this case the `KeyErrror` interrupts the validation phase and so the final message for validation errors is not reported.

Thanks again, and sorry for the confusion!
